<center> <img src = https://raw.githubusercontent.com/AndreyRysistov/DatasetsForPandas/main/hh%20label.jpg alt="drawing" style="width:400px;">

# <center> Проект: Анализ вакансий из HeadHunter
   

In [1]:
import pandas as pd 
import psycopg2
import requests
from bs4 import BeautifulSoup

In [6]:
connection = psycopg2.connect(
    dbname=DBNAME,
    user=USER,
    host=HOST,
    password=PASSWORD,
    port=PORT
)

# Юнит 3. Предварительный анализ данных

1. Напишите запрос, который посчитает количество вакансий в нашей базе (вакансии находятся в таблице vacancies). 

In [4]:
# текст запроса
query_3_1 = f'''
SELECT COUNT(*) AS "Количество вакансий"
FROM vacancies
'''

In [5]:
# результат запроса
df = pd.read_sql_query(query_3_1, connection)
df

C:\Users\lazar\AppData\Local\Temp\ipykernel_11096\1114379196.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_1, connection)


Количество вакансий
0                49197

2. Напишите запрос, который посчитает количество работодателей (таблица employers). 

In [124]:
# текст запроса
query_3_2 = f'''
SELECT COUNT(*) AS "Количество работодателей"
FROM employers
'''

In [125]:
# результат запроса
df = pd.read_sql_query(query_3_2, connection)
df

C:\Users\lazar\AppData\Local\Temp\ipykernel_10964\3062646138.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_2, connection)


Количество работодателей
0                     23501

3. Посчитате с помощью запроса количество регионов (таблица areas).

In [126]:
# текст запроса
query_3_3 = f'''
SELECT COUNT(*) AS "Количество регионов"
FROM areas
'''

In [127]:
# результат запроса
df = pd.read_sql_query(query_3_3, connection)
df

C:\Users\lazar\AppData\Local\Temp\ipykernel_10964\928149064.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_3, connection)


Количество регионов
0                 1362

4. Посчитате с помощью запроса количество сфер деятельности в базе (таблица industries).

In [128]:
# текст запроса
query_3_4 = f'''
SELECT COUNT(*) AS "Количество сфер деятельности"
FROM industries
'''

In [129]:
# результат запроса
df = pd.read_sql_query(query_3_4, connection)
df

C:\Users\lazar\AppData\Local\Temp\ipykernel_10964\2101691108.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_4, connection)


Количество сфер деятельности
0                           294

***

Судя по количеству вакансий и работодателей можно предположить, что у нас в распоряжении стандартная база данных без отклонений, вакансий больше чем работодателей более чем в 2 раза.

# Юнит 4. Детальный анализ вакансий

1. Напишите запрос, который позволит узнать, сколько (cnt) вакансий в каждом регионе (area).
Отсортируйте по количеству вакансий в порядке убывания.

In [132]:
# текст запроса
query_4_1 = f'''
SELECT areas.name AS "Именование региона", COUNT(*) AS "Количество вакансий"
FROM vacancies
JOIN areas ON area_id = areas.id
GROUP BY areas.name
ORDER BY COUNT(*) DESC
'''

In [133]:
# результат запроса
df = pd.read_sql_query(query_4_1, connection)
df

C:\Users\lazar\AppData\Local\Temp\ipykernel_10964\2457173557.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_1, connection)


Именование региона  Количество вакансий
0               Москва                 5333
1      Санкт-Петербург                 2851
2                Минск                 2112
3          Новосибирск                 2006
4               Алматы                 1892
..                 ...                  ...
764         Тарко-Сале                    1
765      Новоаннинский                    1
766              Бирск                    1
767             Сасово                    1
768               Анна                    1

[769 rows x 2 columns]

2. Напишите запрос, чтобы определить у какого количества вакансий заполнено хотя бы одно из двух полей с зарплатой.

In [138]:
# текст запроса
query_4_2 = f'''
SELECT COUNT(*) AS "Количество удовлетворяющих вакансий"
FROM vacancies
WHERE salary_from IS NOT NULL or salary_to IS NOT NULL
'''

In [139]:
# результат запроса
df = pd.read_sql_query(query_4_2, connection)
df

C:\Users\lazar\AppData\Local\Temp\ipykernel_10964\970655580.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_2, connection)


Количество удовлетворяющих вакан
0                             24073

3. Найдите средние значения для нижней и верхней границы зарплатной вилки. Округлите значения до целого.

In [142]:
# текст запроса
query_4_3 = f'''
SELECT ROUND(AVG(salary_from)) AS "Нижняя граница", ROUND(AVG(salary_to)) AS "Верхняя граница"
FROM vacancies
'''

In [143]:
# результат запроса
df = pd.read_sql_query(query_4_3, connection)
df

C:\Users\lazar\AppData\Local\Temp\ipykernel_10964\2500607024.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_3, connection)


Нижняя граница  Верхняя граница
0         71065.0         110537.0

4. Напишите запрос, который выведет количество вакансий для каждого сочетания типа рабочего графика (schedule) и типа трудоустройства (employment), используемого в вакансиях. Результат отсортируйте по убыванию количества.


In [144]:
# текст запроса
query_4_4 = f'''
SELECT schedule AS "Тип графика", employment AS "Тип трудоустройства", COUNT(*) AS "Количество"
FROM vacancies
GROUP BY schedule, employment
ORDER BY COUNT(*) DESC
'''

In [145]:
# результат запроса
df = pd.read_sql_query(query_4_4, connection)
df

C:\Users\lazar\AppData\Local\Temp\ipykernel_10964\2022935265.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_4, connection)


Тип графика  Тип трудоустройства  Количество
0        Полный день     Полная занятость       35367
1   Удаленная работа     Полная занятость        7802
2      Гибкий график     Полная занятость        1593
3   Удаленная работа  Частичная занятость        1312
4     Сменный график     Полная занятость         940
5        Полный день           Стажировка         569
6     Вахтовый метод     Полная занятость         367
7        Полный день  Частичная занятость         347
8      Гибкий график  Частичная занятость         312
9        Полный день     Проектная работа         141
10  Удаленная работа     Проектная работа         133
11     Гибкий график           Стажировка         116
12    Сменный график  Частичная занятость         101
13  Удаленная работа           Стажировка          64
14     Гибкий график     Проектная работа          18
15    Сменный график           Стажировка          12
16    Вахтовый метод     Проектная работа           2
17    Сменный график     Проектная работа           1

5. Напишите запрос, выводящий значения поля Требуемый опыт работы (experience) в порядке возрастания количества вакансий, в которых указан данный вариант опыта. 

In [146]:
# текст запроса
query_4_5 = f'''
SELECT experience AS "Опыт работы", COUNT(*) AS "Количество"
FROM vacancies
GROUP BY experience
ORDER BY COUNT(*)
'''

In [147]:
# результат запроса
df = pd.read_sql_query(query_4_5, connection)
df

C:\Users\lazar\AppData\Local\Temp\ipykernel_10964\2316260449.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_5, connection)


Опыт работы  Количество
0         Более 6 лет        1337
1           Нет опыта        7197
2       От 3 до 6 лет       14511
3  От 1 года до 3 лет       26152

***

Исходя из данных запросов к базе данных, можно обратить внимание на то, что мы с больший долей вероятности наблюдаем стандартное распределение вакансиий, обычно вакансий всегда больше где нет опыта работы, либо около того, в нашем случае большая часть вакансий занята категориями без опыта либо от 3 до 6 лет. В любом случае на первый взгляд явных отклонений мы не заметили, здесь нужен более детальный анализ.

# Юнит 5. Анализ работодателей

1. Напишите запрос, который позволит узнать, какие работодатели находятся на первом и пятом месте по количеству вакансий.

In [148]:
# текст запроса
query_5_1 = f'''
SELECT employers.name AS "Наименование работодателя", COUNT(*) AS "Количество"
FROM vacancies
JOIN employers ON employer_id = employers.id
GROUP BY employers.name
ORDER BY COUNT(*) DESC
'''

In [149]:
# результат запроса
df = pd.read_sql_query(query_5_1, connection)
df

C:\Users\lazar\AppData\Local\Temp\ipykernel_10964\3834399874.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_1, connection)


Наименование работодателя  Количество
0                        Яндекс        1933
1                    Ростелеком         491
2                      Тинькофф         444
3                          СБЕР         428
4                 Газпром нефть         331
...                         ...         ...
14761              им. Тьюринга           1
14762          LLC Novalab Tech           1
14763     Компания Мир упаковки           1
14764    Лифинцев Олег Иванович           1
14765        ТрансЭнергоХолдинг           1

[14766 rows x 2 columns]

2. Напишите запрос, который для каждого региона выведет количество работодателей и вакансий в нём.
Среди регионов, в которых нет вакансий, найдите тот, в котором наибольшее количество работодателей.


In [150]:
# текст запроса
query_5_2 = f'''
SELECT areas.name, COUNT(employers.area) AS "Количество работодателей", COUNT(vacancies.area_id) AS "Количество вакансий"
FROM areas
LEFT JOIN employers ON areas.id = employers.area
LEFT JOIN vacancies ON areas.id = vacancies.area_id
GROUP BY areas.name
HAVING COUNT(vacancies.area_id) = 0
ORDER BY COUNT(employers.area) DESC
LIMIT 1
'''

In [151]:
# результат запроса
df = pd.read_sql_query(query_5_2, connection)
df

C:\Users\lazar\AppData\Local\Temp\ipykernel_10964\2250345976.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_2, connection)


name  Количество работодателей  Количество вакансий
0  Россия                       410                    0

3. Для каждого работодателя посчитайте количество регионов, в которых он публикует свои вакансии. Отсортируйте результат по убыванию количества.


In [152]:
# текст запроса
query_5_3 = f'''
SELECT employers.name AS "Именование работодателя", COUNT(DISTINCT area_id) AS "Количество регионов с вакансиями работодателей"
FROM employers
LEFT JOIN vacancies ON vacancies.employer_id = employers.id
GROUP BY employers.name
ORDER BY COUNT(area_id) DESC
'''

In [153]:
# результат запроса
df = pd.read_sql_query(query_5_3, connection)
df

C:\Users\lazar\AppData\Local\Temp\ipykernel_10964\201667002.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_3, connection)


Именование работодателя  \
0                                               Яндекс   
1                                           Ростелеком   
2                                             Тинькофф   
3                                                 СБЕР   
4                                        Газпром нефть   
...                                                ...   
23170  Независимое агентство Аналитика товарных рынков   
23171             Независимое Информационное Агентство   
23172              Независимое телевидение Севастополя   
23173                                    Независимость   
23174                                             ICNX   

       Количество регионов с вакансиями   
0                                    181  
1                                    152  
2                                     43  
3                                     24  
4                                     24  
...                                  ...  
23170                                  0  
23171                                  0  
23172                                  0  
23173                                  0  
23174                                  0  

[23175 rows x 2 columns]

4. Напишите запрос для подсчёта количества работодателей, у которых не указана сфера деятельности. 

In [182]:
# текст запроса
query_5_4 = f'''
SELECT COUNT(*) AS "Работодателей без сферы"
FROM employers
LEFT JOIN employers_industries ON employer_id = employers.id
WHERE employer_id IS NULL
'''

In [183]:
# результат запроса
df = pd.read_sql_query(query_5_4, connection)
df

C:\Users\lazar\AppData\Local\Temp\ipykernel_10964\3493105328.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_4, connection)


Работодателей без сферы
0                     8419

5. Напишите запрос, чтобы узнать название компании, находящейся на третьем месте в алфавитном списке (по названию) компаний, у которых указано четыре сферы деятельности. 

In [178]:
# текст запроса
query_5_5 = f'''
SELECT name AS "Наименование компании"
FROM employers
JOIN employers_industries ON employer_id = employers.id
GROUP BY name
HAVING COUNT(employer_id) = 4
ORDER BY name
OFFSET 2
LIMIT 1
'''

In [179]:
# результат запроса
df = pd.read_sql_query(query_5_5, connection)
df

C:\Users\lazar\AppData\Local\Temp\ipykernel_10964\388711819.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_5, connection)


Наименование компании
0                  2ГИС

6. С помощью запроса выясните, у какого количества работодателей в качестве сферы деятельности указана Разработка программного обеспечения.


In [192]:
# текст запроса
query_5_6 = f'''
SELECT industries.name AS "Сфера деятельности", COUNT(*) AS "Количество"
FROM employers
JOIN employers_industries ON employer_id = employers.id
JOIN industries ON industries.id = industry_id
WHERE industries.name = 'Разработка программного обеспечения'
GROUP BY industries.name
'''

In [193]:
# результат запроса
df = pd.read_sql_query(query_5_6, connection)
df

C:\Users\lazar\AppData\Local\Temp\ipykernel_10964\821628714.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_6, connection)


Сфера деятельности  Количество
0  Разработка программного обеспечения        3553

7. Для компании «Яндекс» выведите список регионов-миллионников, в которых представлены вакансии компании, вместе с количеством вакансий в этих регионах. Также добавьте строку Total с общим количеством вакансий компании. Результат отсортируйте по возрастанию количества.

Список городов-милионников надо взять [отсюда](https://ru.wikipedia.org/wiki/%D0%93%D0%BE%D1%80%D0%BE%D0%B4%D0%B0-%D0%BC%D0%B8%D0%BB%D0%BB%D0%B8%D0%BE%D0%BD%D0%B5%D1%80%D1%8B_%D0%A0%D0%BE%D1%81%D1%81%D0%B8%D0%B8). 

Если возникнут трудности с этим задание посмотрите материалы модуля  PYTHON-17. Как получать данные из веб-источников и API. 

In [10]:
# код для получения списка городов-милионников
url = 'https://ru.wikipedia.org/wiki/%D0%93%D0%BE%D1%80%D0%BE%D0%B4%D0%B0-%D0%BC%D0%B8%D0%BB%D0%BB%D0%B8%D0%BE%D0%BD%D0%B5%D1%80%D1%8B_%D0%A0%D0%BE%D1%81%D1%81%D0%B8%D0%B8' # Определяем значение URL страницы для запроса
response = requests.get(url) # Делаем GET-запрос к ресурсу и результат ответа сохраняем в переменной response
page = BeautifulSoup(response.text, 'html.parser') # Создаём объект BeautifulSoup, указывая html-парсер
city_list = list()
for tr_row in page.table.tbody:
    city_list.append(tr_row.a.text)
city_list = city_list[1:]
print(city_list)

['Москва', 'Санкт-Петербург', 'Новосибирск', 'Екатеринбург', 'Казань', 'Красноярск', 'Нижний Новгород', 'Челябинск', 'Уфа', 'Краснодар', 'Самара', 'Ростов-на-Дону', 'Омск', 'Воронеж', 'Пермь', 'Волгоград']


In [11]:
# текст запроса
cities_str = ", ".join(f"'{city}'" for city in city_list)
query_5_7 = f'''
SELECT areas.name AS "Наименование региона", COUNT(*) AS "Количество вакансий"
FROM areas
JOIN vacancies ON area_id = areas.id
JOIN employers ON employers.id = vacancies.employer_id
WHERE areas.name IN ({cities_str}) AND employers.name = 'Яндекс' 
GROUP BY areas.name

UNION ALL
SELECT 'total', COUNT(*)
FROM areas
JOIN vacancies ON area_id = areas.id
JOIN employers ON employers.id = vacancies.employer_id
WHERE areas.name IN ({cities_str}) AND employers.name = 'Яндекс'
ORDER BY 2
'''


In [17]:
# результат запроса
df = pd.read_sql_query(query_5_7, connection)
df

C:\Users\lazar\AppData\Local\Temp\ipykernel_12804\1965596796.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_7, connection)


Наименование региона  Количество вакансий
0                  Омск                   21
1             Челябинск                   22
2            Красноярск                   23
3             Волгоград                   24
4                 Пермь                   25
5                Казань                   25
6        Ростов-на-Дону                   25
7                   Уфа                   26
8                Самара                   26
9             Краснодар                   30
10              Воронеж                   32
11          Новосибирск                   35
12      Нижний Новгород                   36
13         Екатеринбург                   39
14      Санкт-Петербург                   42
15               Москва                   54
16                total                  485

***

Среди всех работодателей, большая часть вакансий выставляет компания, именуемая как "Яндекс", также в одном из запросов высветился регион, где находится больше всего работодателей, но без вакансий - "Россия", что мне кажется не корректным.


# Юнит 6. Предметный анализ

1. Сколько вакансий имеет отношение к данным?

Считаем, что вакансия имеет отношение к данным, если в её названии содержатся слова 'data' или 'данн'.

*Подсказка: Обратите внимание, что названия вакансий могут быть написаны в любом регистре.* 


In [45]:
# текст запроса
query_6_1 = f'''
SELECT COUNT(*) AS "Вакансии data"
FROM vacancies
WHERE lower(name) LIKE '%data%' OR lower(name) LIKE '%данн%'
'''

In [46]:
# результат запроса
df = pd.read_sql_query(query_6_1, connection)
df

C:\Users\lazar\AppData\Local\Temp\ipykernel_12804\366702258.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_1, connection)


Вакансии data
0           1771

2. Сколько есть подходящих вакансий для начинающего дата-сайентиста? 
Будем считать вакансиями для дата-сайентистов такие, в названии которых есть хотя бы одно из следующих сочетаний:
* 'data scientist'
* 'data science'
* 'исследователь данных'
* 'ML' (здесь не нужно брать вакансии по HTML)
* 'machine learning'
* 'машинн%обучен%'

** В следующих заданиях мы продолжим работать с вакансиями по этому условию.*

Считаем вакансиями для специалистов уровня Junior следующие:
* в названии есть слово 'junior' *или*
* требуемый опыт — Нет опыта *или*
* тип трудоустройства — Стажировка.
 

In [60]:
# текст запроса
query_6_2 = f'''
SELECT COUNT(*) AS "Количество junior вакансий"
FROM vacancies
WHERE
    ((lower(name) LIKE '%data scientist%' OR
    lower(name) LIKE '%data science%' OR
    lower(name) LIKE '%исследователь данных%' OR
    (lower(name) LIKE '%ml%' AND lower(name) NOT LIKE'%html%') OR
    lower(name) LIKE '%machine learning%' OR
    lower(name) LIKE '%машинн%обучен%'))
    AND
    (lower(name) LIKE '%junior%' OR
    experience LIKE 'Нет опыта' OR
    employment LIKE 'Стажировка')
'''

In [61]:
# результат запроса
df = pd.read_sql_query(query_6_2, connection)
df

C:\Users\lazar\AppData\Local\Temp\ipykernel_11096\2267038862.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_2, connection)


Количество junior вакансий
0                          51

3. Сколько есть вакансий для DS, в которых в качестве ключевого навыка указан SQL или postgres?

** Критерии для отнесения вакансии к DS указаны в предыдущем задании.*

In [66]:
# текст запроса
query_6_3 = f'''
SELECT COUNT(*) AS "Вакансий с SQL или postgres"
FROM vacancies
WHERE
    ((lower(name) LIKE '%data scientist%' OR
    lower(name) LIKE '%data science%' OR
    lower(name) LIKE '%исследователь данных%' OR
    (lower(name) LIKE '%ml%' AND lower(name) NOT LIKE'%html%') OR
    lower(name) LIKE '%machine learning%' OR
    lower(name) LIKE '%машинн%обучен%'))
    AND
    (lower(key_skills) LIKE '%sql%' OR
    lower(key_skills) LIKE '%postgres%')
'''

In [67]:
# результат запроса
df = pd.read_sql_query(query_6_3, connection)
df

C:\Users\lazar\AppData\Local\Temp\ipykernel_11096\3794299702.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_3, connection)


Вакансий с SQL или postgres
0                          229

4. Проверьте, насколько популярен Python в требованиях работодателей к DS.Для этого вычислите количество вакансий, в которых в качестве ключевого навыка указан Python.

** Это можно сделать помощью запроса, аналогичного предыдущему.*

In [68]:
# текст запроса
query_6_4 = f'''
SELECT COUNT(*) AS "Количество вакансий с python"
FROM vacancies
WHERE
    ((lower(name) LIKE '%data scientist%' OR
    lower(name) LIKE '%data science%' OR
    lower(name) LIKE '%исследователь данных%' OR
    (lower(name) LIKE '%ml%' AND lower(name) NOT LIKE'%html%') OR
    lower(name) LIKE '%machine learning%' OR
    lower(name) LIKE '%машинн%обучен%'))
    AND
    (lower(key_skills) LIKE '%python%')
'''

In [69]:
# результат запроса
df = pd.read_sql_query(query_6_4, connection)
df

C:\Users\lazar\AppData\Local\Temp\ipykernel_11096\4264665179.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_4, connection)


Количество вакансий с python
0                           357

5. Сколько ключевых навыков в среднем указывают в вакансиях для DS?
Ответ округлите до двух знаков после точки-разделителя.

In [58]:
# текст запроса
query_6_5 = f'''
SELECT ROUND(AVG(CHAR_LENGTH(key_skills) - CHAR_LENGTH(replace(key_skills, CHR(9), '')) + 1), 2) AS "Среднее количество навыков"

FROM vacancies
WHERE
    (lower(name) LIKE '%data scientist%' OR
    lower(name) LIKE '%data science%' OR
    lower(name) LIKE '%исследователь данных%' OR
    (lower(name) LIKE '%ml%' AND lower(name) NOT LIKE'%html%') OR
    lower(name) LIKE '%machine learning%' OR
    lower(name) LIKE '%машинн%обучен%') AND
    (key_skills IS NOT NULL AND key_skills != '')
'''

In [59]:
# результат запроса
df = pd.read_sql_query(query_6_5, connection)
df

C:\Users\lazar\AppData\Local\Temp\ipykernel_11096\3654210517.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_5, connection)


Среднее количество навыков
0                        6.55

6. Напишите запрос, позволяющий вычислить, какую зарплату для DS в **среднем** указывают для каждого типа требуемого опыта (уникальное значение из поля *experience*). 

При решении задачи примите во внимание следующее:
1. Рассматриваем только вакансии, у которых заполнено хотя бы одно из двух полей с зарплатой.
2. Если заполнены оба поля с зарплатой, то считаем зарплату по каждой вакансии как сумму двух полей, делённую на 2. Если заполнено только одно из полей, то его и считаем зарплатой по вакансии.
3. Если в расчётах участвует null, в результате он тоже даст null (посмотрите, что возвращает запрос select 1 + null). Чтобы избежать этой ситуацию, мы воспользуемся функцией [coalesce](https://postgrespro.ru/docs/postgresql/9.5/functions-conditional#functions-coalesce-nvl-ifnull), которая заменит null на значение, которое мы передадим. Например, посмотрите, что возвращает запрос `select 1 + coalesce(null, 0)`

Выясните, на какую зарплату в среднем может рассчитывать дата-сайентист с опытом работы от 3 до 6 лет. Результат округлите до целого числа. 

In [70]:
# текст запроса
query_6_6 = f'''
SELECT experience AS "Опыт", ROUND(AVG("Средняя зарплата")) AS "Средняя зарплата по опыту"
FROM (
SELECT experience, (salary_from + salary_to) / 2 AS "Средняя зарплата"
FROM vacancies
WHERE
    (lower(name) LIKE '%data scientist%' OR
    lower(name) LIKE '%data science%' OR
    lower(name) LIKE '%исследователь данных%' OR
    (lower(name) LIKE '%ml%' AND lower(name) NOT LIKE'%html%') OR
    lower(name) LIKE '%machine learning%' OR
    lower(name) LIKE '%машинн%обучен%') AND
    (salary_from IS NOT NULL AND salary_to IS NOT NULL)
UNION ALL
SELECT experience, salary_from
FROM vacancies
WHERE
    (lower(name) LIKE '%data scientist%' OR
    lower(name) LIKE '%data science%' OR
    lower(name) LIKE '%исследователь данных%' OR
    (lower(name) LIKE '%ml%' AND lower(name) NOT LIKE'%html%') OR
    lower(name) LIKE '%machine learning%' OR
    lower(name) LIKE '%машинн%обучен%') AND
    (salary_from IS NOT NULL AND salary_to IS NULL)
UNION ALL
SELECT experience, salary_to
FROM vacancies
WHERE
    (lower(name) LIKE '%data scientist%' OR
    lower(name) LIKE '%data science%' OR
    lower(name) LIKE '%исследователь данных%' OR
    (lower(name) LIKE '%ml%' AND lower(name) NOT LIKE'%html%') OR
    lower(name) LIKE '%machine learning%' OR
    lower(name) LIKE '%машинн%обучен%') AND
    (salary_from IS NULL AND salary_to IS NOT NULL)
) AS combined_data
GROUP BY experience
ORDER BY experience
'''

In [71]:
# результат запроса
df = pd.read_sql_query(query_6_6, connection)
df

C:\Users\lazar\AppData\Local\Temp\ipykernel_11096\519194770.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_6, connection)


Опыт  Средняя зарплата по опыту
0         Более 6 лет                   157933.0
1           Нет опыта                    74643.0
2  От 1 года до 3 лет                   144233.0
3       От 3 до 6 лет                   256454.0

***

В ходе предметного анализа, было замечено, что наибольшую среднюю заработную плату в вакансиях выставляют людям с опытом от 3 до 6 лет, что мне кажется не совсем корректным так как у нас ещё есть опыт более 6 лет, по идее такой уровень должен ценится больше на рынке труда, но в нашем случае, это говорит о том, что в данной категории мало вакансий, а также работодатели склонны принимать решение о не выставлении заработной платы в данной категории. Также было замечено, что вакансий по data science для уровня junior очень мало.

# Общий вывод по проекту

В ходе работы над проектом был замечен ряд тенденций и решений работодателей которые мешают в той или иной степени найму людей в свои компании.

Первое, на что хотелось бы обратить внимание это количество работодателей без сферы деятельности, их количество составляет 8419, из общего числа 23501, это более 1/3 части от всего числа работодателей. Искатели как правило интересуются и ищут не только по вакансиям, но и по сфере деятельности работодателей, чтобы понять подойдут ли ключевые навыки под сферу деятельности компании, упрощая сказанное искатель мог просто не найти работодателя.

Второе, был выявлен "регион" в котором указаны 410 работодателей, а вакансий 0. В поле регион высветилась "Россия", это страна, не регион в общепринятом смысле, также HeadHunter (hh.ru) – российская компания, которая ориентируется в большей степени на работодателях из России, Белорусии и Казахстана. Это может сигнализировать о том, что работодатель скорее всего выставляет вакансии с отдельным регионом страны, что логически может попадать под Россию, но системой это не учитывается.

In [18]:
query_7_1 = f'''
SELECT areas.name AS "Наименование региона", COUNT(*) AS "Количество junior вакансий"
FROM areas
JOIN vacancies ON area_id = areas.id
JOIN employers ON employers.id = vacancies.employer_id
WHERE 
    ((lower(vacancies.name) LIKE '%data scientist%' OR
    lower(vacancies.name) LIKE '%data science%' OR
    lower(vacancies.name) LIKE '%исследователь данных%' OR
    (lower(vacancies.name) LIKE '%ml%' AND lower(vacancies.name) NOT LIKE'%html%') OR
    lower(vacancies.name) LIKE '%machine learning%' OR
    lower(vacancies.name) LIKE '%машинн%обучен%'))
    AND
    (lower(vacancies.name) LIKE '%junior%' OR
    experience LIKE 'Нет опыта' OR
    employment LIKE 'Стажировка')
    AND
    (areas.name IN ({cities_str}))
GROUP BY areas.name
ORDER BY 2
'''
# результат запроса
df = pd.read_sql_query(query_7_1, connection)
df

C:\Users\lazar\AppData\Local\Temp\ipykernel_13236\2792894179.py:23: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_7_1, connection)


Наименование региона  Количество junior вакансий
0      Нижний Новгород                           1
1                 Омск                           1
2               Казань                           2
3                Пермь                           2
4              Воронеж                           2
5       Ростов-на-Дону                           2
6         Екатеринбург                           2
7          Новосибирск                           5
8      Санкт-Петербург                           9
9               Москва                          19

In [22]:
query_7_2 = f'''
SELECT areas.name AS "Наименование региона", COUNT(*) AS "Количество вакансий"
FROM areas
JOIN vacancies ON area_id = areas.id
JOIN employers ON employers.id = vacancies.employer_id
WHERE 
    ((lower(vacancies.name) LIKE '%data scientist%' OR
    lower(vacancies.name) LIKE '%data science%' OR
    lower(vacancies.name) LIKE '%исследователь данных%' OR
    (lower(vacancies.name) LIKE '%ml%' AND lower(vacancies.name) NOT LIKE'%html%') OR
    lower(vacancies.name) LIKE '%machine learning%' OR
    lower(vacancies.name) LIKE '%машинн%обучен%'))
    AND
    (areas.name IN ({cities_str}))
GROUP BY areas.name
ORDER BY 2
'''
# результат запроса
df = pd.read_sql_query(query_7_2, connection)
df

C:\Users\lazar\AppData\Local\Temp\ipykernel_13236\4088367907.py:19: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_7_2, connection)


Наименование региона  Количество вакансий
0             Волгоград                    1
1            Красноярск                    1
2             Челябинск                    1
3                  Омск                    3
4                   Уфа                    4
5               Воронеж                    4
6                Самара                    6
7          Екатеринбург                    7
8             Краснодар                    7
9                 Пермь                    8
10       Ростов-на-Дону                   11
11               Казань                   19
12      Нижний Новгород                   20
13          Новосибирск                   23
14      Санкт-Петербург                   69
15               Москва                  225

В третьих, было замечено, что даже в крупных городах миллионниках вакансий на data science уровня junior очень мало, если брать во внимание наши критерии отбора, рассматривая общее количество вакансий имеющих отношение к data sciense в каждом городе в отдельности к числу junior вакансий о городам, то это разница велика.